keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,6 +1076,335 @@ the remaining requests in that queue start getting faster service.  In
 both cases, the service delivery in the virtual world has reacted
 properly to the true service duration.
 
+### Support for LIST requests
+
+Up until now, we were assuming that even though the requests aren't
+necessarily equally expensive, their actual cost is actually greatly
+reflected by the time it took to process them.  But while being processed
+each of them is consuming the equal amount of resources.
+
+It works well for requests that are touching only a single object.
+However, given the fact that in practise the concurrency limits has to be
+set much higher than number of available cores to achieve reasonable system
+throughput, this no longer works that well for LIST requests that are orders
+of magnitude more expensive.  There are two aspects of that:
+- for CPU the hand-wavy way of rationalizing it is that he ratio of time
+  the request is processed by the processor to the actual time of processing
+  the request starts to visibly differ (e.g. due to I/O waiting time -
+  there is communication with etcd in between for example).
+- for memory the reasoning is more obvious as we simply keep all elements
+  that we process in memory
+
+As a result, kube-apiserver (and etcd) may be able to easily keep with N
+simple in-flight requests (e.g. create or get a single Pod), but will explode
+trying to process N requests listing all the pods in the system at the same
+time.
+
+#### Width of the request
+
+In order to address this problem, we are introducing the concept of `width`
+of the request.  Instead of saying that every request is consuming a single
+unit of concurrency, we allow for a request to consume `<width>` units of
+concurrency while being processed.
+
+This basically means, that the cost of processing a given request is no
+longer reflected by its `<processing latency>` and instead its cost is now
+equal to `<width> x <processing latency>`.  The rationale behind it is that
+the request is now consuming `<width>` concurrency units for the duration
+of its processing.
+
+While in theory the `width` can be an arbitrary non-integer number, for
+practical reasons, we will assume it actually is an integer.  Given that
+our estimations here are very rough anyway that seems a reasonable
+simplification that makes dispatching the budget a bit simpler.
+
+#### Determining the width
+
+While one can imagine arbitrarily sophisticated algorithms for it (including
+exposing defining the width of requests via FlowSchema API), we want to start
+with something relatively simple to first get operational experience with it
+before investing into sophisticated algorithms or exposing a knob to users.
+
+In order to determine the function that will be approximating the `width` of
```

Comment:
@lavalamp - as mentioned in the other comment - I rewrote this section. PTAL if this makes more sense to you now.